grakn-test/src/test/java/ai/grakn/test/engine/lock/ZookeeperLockTest.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Grakn - A Distributed Semantic Database
+ * Copyright (C) 2016  Grakn Labs Limited
+ *
+ * Grakn is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Grakn is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Grakn. If not, see <http://www.gnu.org/licenses/gpl.txt>.
+ */
+
+package ai.grakn.test.engine.lock;
+
+import ai.grakn.engine.lock.ZookeeperLock;
+import ai.grakn.engine.tasks.manager.ZookeeperConnection;
+import ai.grakn.exception.EngineStorageException;
+import ai.grakn.test.EngineContext;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+public class ZookeeperLockTest {
```

Comment:
You can bypass the different constructors easily. For example, the code below is used to get then cache to be tested. In this case using the cache provider.

```
private ConceptCache getCache(Caches caches){
        switch(caches){
            case DISTRIBUTED:
                return EngineCacheProvider.getCache();
            case STAND_ALONE:
                return EngineCacheStandAlone.getCache();
        }
        throw new RuntimeException("Invalid cache [" + caches + "]");
    }
```

You could do something similar. Startup the distributed engine context so when you can to use the ZK lock just use `LockProvider.get(...)`. When you want to use the re-entrant lock just return the re-erntrant lock instead.

> The two don't entirely behave the same way, though.

They should from the interface perspective. `getLock()` from the user perspective should do the same thing regardless of the implementation. So even though they behave differently under the hood, they implementing the same interface and therefore the functional result should be the same. If the functional result is not the same then we may have an issue.